### PR TITLE
[receiver/purefb] New Pure Storage FlashBlade receiver

### LINKED ISSUE #17528: New component: Pure Storage Receiver for FlashBlade
### **Pure Storage Receiver for OpenTelemetry**

PureStorage has a product called **FlashBlade**. Currently, we export the arrays to vendor-specific backends, but we want to provide a receiver that will allow customers to send data to their backends of choice

It is possible to see the current **FlashBlade OpenMetrics exporter** [here](https://github.com/PureStorage-OpenConnect/pure-fb-openmetrics-exporter).

### Example configuration for the component

```yaml
receivers:
  purefb:
    endpoint: http://172.31.60.207:9491/metrics
    arrays:
    - address: gse-fb01
      auth:
        authenticator: bearertokenauth/fb01
    buckets:
    - address: gse-fb01
      auth:
        authenticator: bearertokenauth/fb01
    clients:
    - address: gse-fb01
      auth:
        authenticator: bearertokenauth/fb01
    file:
    - address: gse-fb01
      auth:
        authenticator: bearertokenauth/fb01
    usage:
    - address: gse-fb01
      auth:
        authenticator: bearertokenauth/fb01
    env: dev
    settings:
      reload_intervals:
        array: 
        buckets: 
        clients: 
        file: 
        usage:
```

### Telemetry data types supported

Metrics

### Is this a vendor-specific component?

- [X] This is a vendor-specific component
- [ ] If this is a vendor-specific component, I am proposing to contribute this as a representative of the vendor.

### Sponsor (optional)

Juraci Paixão Kröhling @jpkrohling

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dgoscn**
**Description:** <Describe what has changed.>
New Pure Storage FlashBlade receiver component

**Link to tracking Issue:** 
Closes #17528

**Testing:** 
Just following the README.md is enough to get the correct results

**Documentation:**
The documentation still in development due to the stability of the receiver

### CODE REVIEW

**Comment by dgoscn** [comment]
cc @jpkrohling

File: receiver/purefbreceiver/config.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	Buckets time.Duration `mapstructure:"buckets"`
+	Clients time.Duration `mapstructure:"clients"`
+	File    time.Duration `mapstructure:"file"`
+	Usage   time.Duration `mapstructure:"usage"`
+}
+
+func (c *Config) Validate() error {
+	var err error
+
+	if len(c.Settings.ReloadIntervals.Array.String()) == 0 {
```
Comment (jpkrohling) [inline receiver/purefbreceiver/config.go:71]: This looks strange. `time.Duration` is actually an integer, so you can just do this:

```suggestion
	if c.Settings.ReloadIntervals.Array == 0 {
```

File: receiver/purefbreceiver/config.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+		err = multierr.Append(err, errors.New("Arrays not provided and is required"))
```
Comment (jpkrohling) [inline receiver/purefbreceiver/config.go:72]: Errors should not start with capital letters in Go. Use something like this:

```suggestion
		err = multierr.Append(err, errors.New("reload interval for 'arrays' must be provided"))
```

**Review by jpkrohling — COMMENTED** [review]
Looks good, but the Validate function needs a small adjustment.

File: receiver/purefbreceiver/config.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	// Env represents the respective environment value valid to scrape
+	Env string `mapstructure:"env"`
+}
+
+type Settings struct {
+	ReloadIntervals *ReloadIntervals `mapstructure:"reload_intervals"`
+}
+
+type ReloadIntervals struct {
+	Array   time.Duration `mapstructure:"array"`
```
Comment (chrroberts-pure) [inline receiver/purefbreceiver/config.go:61]: Based on [The OpenMetrics Endpoint list](https://github.com/PureStorage-OpenConnect/pure-fb-openmetrics-exporter#scraping-endpoints) & [The OpenMetrics collector.go](https://github.com/PureStorage-OpenConnect/pure-fb-openmetrics-exporter/blob/main/internal/openmetrics-exporter/collector.go) there's only Array, Usage, and Clients endpoints on the OpenMetrics exporter.

Comment (dgoscn) [reply receiver/purefbreceiver/config.go:61]: Great @chrroberts-pure. Thanks for the advise

File: receiver/purefbreceiver/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  bearertokenauth/fb01:
+    token: "..."
+
+receivers:
+  purefb:
+    endpoint: http://172.31.60.207:9491/metrics
+
+  purefb/with_custom_intervals:
+    endpoint: http://172.31.60.207:9491/metrics
+    arrays:
```
Comment (jpkrohling) [inline receiver/purefbreceiver/README.md:31]: Should this list all endpoints? The doc should also mention that only endpoints explicitly added here will be scraped.

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Please, fix the conflict on the go.mod file. Once that's done, and once @chrroberts-pure approves, I'll merge this.

File: receiver/purefbreceiver/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# Pure Storage FlashBlade Receiver
+
+| Status                   |                     |
+| ------------------------ |---------------------|
+| Stability                | [in-development]    |
+| Supported pipeline types | metrics             |
+| Distributions            | [contrib]           |
+
+The Pure Storage FlashBlade receiver, receives metrics from Pure Storage internal services hosts.
```
Comment (chrroberts-pure) [inline receiver/purefbreceiver/README.md:9]: The Pure Storage FlashBlade receiver, receives metrics from Pure Storage FlashBlade via the [Pure Storage FlashBlade OpenMetrics Exporter](https://github.com/PureStorage-OpenConnect/pure-fb-openmetrics-exporter)

File: receiver/purefbreceiver/README.md (hunk 4 above)
Comment (chrroberts-pure) [reply receiver/purefbreceiver/README.md:31]: +1

File: receiver/purefbreceiver/config_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		id       component.ID
+		expected component.Config
+	}{
+		{
+			id: component.NewID(typeStr),
+			expected: &Config{
+				Settings: &Settings{
+					ReloadIntervals: &ReloadIntervals{
+						Array:   15 * time.Second,
+						Clients: 15 * time.Second,
```
Comment (chrroberts-pure) [inline receiver/purefbreceiver/config_test.go:44]: Let’s set the default for Clients and Usage to 300. These are very expensive calls

File: receiver/purefbreceiver/factory.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		receiver.WithMetrics(createMetricsReceiver, stability))
+}
+
+func createDefaultConfig() component.Config {
+	return &Config{
+		HTTPClientSettings: confighttp.HTTPClientSettings{},
+		Settings: &Settings{
+			ReloadIntervals: &ReloadIntervals{
+				Array:   15 * time.Second,
+				Clients: 15 * time.Second,
```
Comment (chrroberts-pure) [inline receiver/purefbreceiver/factory.go:49]: Set the defaults for Clients and Usage to 300. These are very expensive calls to the flashblade. The timeout should be set higher to account for this as well

File: receiver/purefbreceiver/testdata/config.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+receivers:
+  purefb:
+    endpoint: http://172.31.60.207:9491/metrics
+
+  purefb/with_custom_intervals:
+    endpoint: http://172.31.60.207:9491/metrics
+    arrays:
+      - address: gse-fb01
```
Comment (chrroberts-pure) [inline receiver/purefbreceiver/testdata/config.yaml:8]: Please remove gse prefix rom these host names as they’re internal. Can be set just to “fb01”

**Review by chrroberts-pure — COMMENTED** [review]
Small nits, but otherwise LGTM

**Comment by dgoscn** [comment]
Hi @jpkrohling and @chrroberts-pure can you please double check the updated files from review? 
Thank you

File: receiver/purefbreceiver/README.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        auth:
+          authenticator: bearertokenauth/fb01
+    clients:
+      - address: fb01
+        auth:
+          authenticator: bearertokenauth/fb01
+    env: dev
+    settings:
+      reload_intervals:
+        array: 10s
```
Comment (jpkrohling) [inline receiver/purefbreceiver/README.md:43]: I think this should have the values closer to the ones users want. If the default is 5 minutes, this should be then, say, 6 minutes. Leaving it 10s kind of encourages users to use that value. You'd be surprised to know that most people will just copy and paste this config straight into their collector's :-)

File: receiver/purefbreceiver/factory.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+		receiver.WithMetrics(createMetricsReceiver, stability))
+}
+
+func createDefaultConfig() component.Config {
+	return &Config{
+		HTTPClientSettings: confighttp.HTTPClientSettings{},
+		Settings: &Settings{
+			ReloadIntervals: &ReloadIntervals{
+				Array:   15 * time.Second,
+				Clients: 300 * time.Second,
```
Comment (jpkrohling) [inline receiver/purefbreceiver/factory.go:49]: 300 Seconds is 5 minutes. I personally find it easier to read `5 * time.Minute`.

File: receiver/purefbreceiver/internal/scraper_test.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		extensions: map[component.ID]component.Component{
+			component.NewIDWithName("bearertokenauth", "fb01"): baExt,
+		},
+	}
+
+	endpoint := "http://example.com"
+	interval := 15 * time.Second
+	cfgs := []ScraperConfig{
+		{
+			Address: "gse-fb01",
```
Comment (jpkrohling) [inline receiver/purefbreceiver/internal/scraper_test.go:54]: This probably needs to be changed as well to just `fb01`.

**Review by jpkrohling — COMMENTED** [review]
Just a couple of small details which wouldn't block this, but it looks like there's something outdated with this PR. The CI says:

```
Replace statements are out of date, please run "make crosslink" and commit the changes in this PR.
```

File: receiver/purefbreceiver/factory.go [hunk 12]
```diff
@@ -45,9 +45,9 @@ func createDefaultConfig() component.Config {
 		HTTPClientSettings: confighttp.HTTPClientSettings{},
 		Settings: &Settings{
 			ReloadIntervals: &ReloadIntervals{
-				Array:   15 * time.Second,
-				Clients: 300 * time.Second,
-				Usage:   300 * time.Second,
+				Array:   5 * time.Minute,
```
Comment (jpkrohling) [inline receiver/purefbreceiver/factory.go:48]: @chrroberts-pure, should this be 5 minutes as well, or should this be kept at 15 seconds?

Comment (chrroberts-pure) [reply receiver/purefbreceiver/factory.go:48]: Array should be set to 15s

File: receiver/purefbreceiver/testdata/config.yaml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        auth:
+          authenticator: bearertokenauth/fb01
+    usage:
+      - address: fb01
+        auth:
+          authenticator: bearertokenauth/fb01
+    env: dev
+    settings:
+      reload_intervals:
+        array: 5m
```
Comment (chrroberts-pure) [inline receiver/purefbreceiver/testdata/config.yaml:22]: 30s here should be a good test

**Review by jpkrohling — APPROVED** [review]

File: receiver/purefbreceiver/go.mod [hunk 14]
```diff
@@ -0,0 +1,182 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/receiver/purefbreceiver
+
+go 1.18
+
+require (
+	github.com/open-telemetry/opentelemetry-collector-contrib/extension/bearertokenauthextension v0.69.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver v0.69.0
```
Comment (jpkrohling) [inline receiver/purefbreceiver/go.mod:7]:
```suggestion
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver v0.70.0
```

File: receiver/purefbreceiver/go.mod [hunk 15]
```diff
@@ -0,0 +1,182 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/receiver/purefbreceiver
+
+go 1.18
+
+require (
+	github.com/open-telemetry/opentelemetry-collector-contrib/extension/bearertokenauthextension v0.69.0
```
Comment (jpkrohling) [inline receiver/purefbreceiver/go.mod:6]: There was a new release this morning, this should be:

```suggestion
	github.com/open-telemetry/opentelemetry-collector-contrib/extension/bearertokenauthextension v0.70.0
```

**Review by jpkrohling — APPROVED** [review]
Just found another issue, but this can be fixed on a follow-up PR if CI allows.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**